### PR TITLE
Webpacker

### PR DESCRIPTION
**Pull request by gmq**
Add support for ActiveAdmin in webpacker mode. (#298)

- Uses rollup to generate the javascript used by sprockets
- Updates javascript files to use es6 imports.
- Adds generators to include the appropriate files when installing depending on how ActiveAdmin was installed.

TODO:
- It assumes the npm package is going to be named "activeadmin_addons" How are we going to handle that release?
- ~~Tests~~
- Potassium support will be handled separately.

For testing:
- Clone this branch
- Run `yarn prepublishOnly` inside cloned folder
- In a new Rails installation add activeadmin and add this gem locally
`gem 'activeadmin_addons', path: '../activeadmin_addons'`
- Install activeadmin in webpacker mode
`bundle exec rails g active_admin:install --use_webpacker`
- Install activeadmin addons
`bundle exec rails g activeadmin_addons:install`
(ignore yarn error)
- Install npm package
`yarn add ../activeadmin_addons`

### CODE REVIEW

File: rollup.config.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    commonjs(),
+    babel(),
+    uglify(uglifyOptions),
+  ],
+  // Use client's yarn dependencies instead of bundling everything
+  external: [
+    'jquery-datetimepicker',
+    'select2',
+  ]
+}
```
Comment (monkeyci) [inline rollup.config.js:41]: Missing semicolon                                semi<br>Newline required at end of file but not found    eol-last

File: rollup.config.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    resolve(),
+    commonjs(),
+    babel(),
+    uglify(uglifyOptions),
+  ],
+  // Use client's yarn dependencies instead of bundling everything
+  external: [
+    'jquery-datetimepicker',
+    'select2',
+  ]
```
Comment (monkeyci) [inline rollup.config.js:40]: Missing trailing comma                           comma-dangle

File: rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    ` + '\n'
+  }
+}
+
+export default {
+  input: "app/javascript/activeadmin_addons/all.js",
+  output: {
+    file: "app/assets/javascripts/activeadmin_addons/all.js",
+    format: "umd",
+    name: "ActiveAdmin Addons"
```
Comment (monkeyci) [inline rollup.config.js:28]: Strings must use singlequote                     quotes<br>Missing trailing comma                           comma-dangle

File: rollup.config.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      //= require jquery.xdan.datetimepicker.full
+    ` + '\n'
+  }
+}
+
+export default {
+  input: "app/javascript/activeadmin_addons/all.js",
+  output: {
+    file: "app/assets/javascripts/activeadmin_addons/all.js",
+    format: "umd",
```
Comment (monkeyci) [inline rollup.config.js:27]: Strings must use singlequote                     quotes

File: rollup.config.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      //= require select2.full
+      //= require jquery.xdan.datetimepicker.full
+    ` + '\n'
+  }
+}
+
+export default {
+  input: "app/javascript/activeadmin_addons/all.js",
+  output: {
+    file: "app/assets/javascripts/activeadmin_addons/all.js",
```
Comment (monkeyci) [inline rollup.config.js:26]: Strings must use singlequote                     quotes

File: rollup.config.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+       * Warning: This file is auto-generated, do not modify. Instead, make your changes in 'app/javascript/activeadmin_addons/' and run \`yarn build\`
+       */
+      //= require select2.full
+      //= require jquery.xdan.datetimepicker.full
+    ` + '\n'
+  }
+}
+
+export default {
+  input: "app/javascript/activeadmin_addons/all.js",
```
Comment (monkeyci) [inline rollup.config.js:24]: Strings must use singlequote                     quotes

File: rollup.config.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    indent_level: 2,
+    preamble: stripIndent`
+      /**
+       * Warning: This file is auto-generated, do not modify. Instead, make your changes in 'app/javascript/activeadmin_addons/' and run \`yarn build\`
+       */
+      //= require select2.full
+      //= require jquery.xdan.datetimepicker.full
+    ` + '\n'
+  }
+}
```
Comment (monkeyci) [inline rollup.config.js:21]: Missing semicolon                                semi

File: rollup.config.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    beautify: true,
+    indent_level: 2,
+    preamble: stripIndent`
+      /**
+       * Warning: This file is auto-generated, do not modify. Instead, make your changes in 'app/javascript/activeadmin_addons/' and run \`yarn build\`
+       */
+      //= require select2.full
+      //= require jquery.xdan.datetimepicker.full
+    ` + '\n'
+  }
```
Comment (monkeyci) [inline rollup.config.js:20]: Missing trailing comma                           comma-dangle

File: rollup.config.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  output: {
+    beautify: true,
+    indent_level: 2,
+    preamble: stripIndent`
+      /**
+       * Warning: This file is auto-generated, do not modify. Instead, make your changes in 'app/javascript/activeadmin_addons/' and run \`yarn build\`
+       */
+      //= require select2.full
+      //= require jquery.xdan.datetimepicker.full
+    ` + '\n'
```
Comment (monkeyci) [inline rollup.config.js:19]: Missing trailing comma                           comma-dangle

File: rollup.config.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+import babel from "rollup-plugin-babel";
+import { uglify } from "rollup-plugin-uglify";
+import { stripIndent } from 'common-tags';
+
+const uglifyOptions = {
+  mangle: false,
+  compress: false,
+  output: {
+    beautify: true,
+    indent_level: 2,
```
Comment (monkeyci) [inline rollup.config.js:12]: Identifier 'indent_level' is not in camel case   camelcase

File: rollup.config.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import resolve from 'rollup-plugin-node-resolve';
```
Comment (monkeyci) [inline rollup.config.js:1]: Definition for rule 'vue/max-len' was not found  vue/max-len

File: lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        to_add = "import \"activeadmin_addons\"\n"
+        to_add
+      end
+
+      def css_assets
+        "@import 'activeadmin_addons/src/stylesheets/all';\n" 
+      end
+    end
+  end
+end
```
Comment (monkeyci) [inline lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb:35]: Final newline missing.

File: lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+      private
+
+      def js_assets
+        to_add = "import \"activeadmin_addons\"\n"
+        to_add
+      end
+
+      def css_assets
+        "@import 'activeadmin_addons/src/stylesheets/all';\n"
```
Comment (monkeyci) [inline lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb:31]: Trailing whitespace detected.

File: lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        reference = "import \"@activeadmin/activeadmin\";\n"
+
+        inject_into_file("#{file_path}", js_assets, after: reference)
+      end
+
+      def add_stylesheets
+        file_path = 'app/javascript/stylesheets/active_admin.scss'
+
+        begin
+          prepend_file("#{file_path}", css_assets)
```
Comment (monkeyci) [inline lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb:15]: Prefer to_s over string interpolation.

File: lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb [hunk 15]
```diff
@@ -0,0 +1,35 @@
+module ActiveadminAddons
+  module Generators
+    class WebpackerGenerator < Rails::Generators::Base
+      def add_javascripts
+        file_path = 'app/javascript/packs/active_admin.js'
+        reference = "import \"@activeadmin/activeadmin\";\n"
+
+        inject_into_file("#{file_path}", js_assets, after: reference)
```
Comment (monkeyci) [inline lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb:8]: Prefer to_s over string interpolation.

File: lib/generators/activeadmin_addons/install/install_generator.rb [hunk 16]
```diff
@@ -2,55 +2,16 @@ module ActiveadminAddons
   module Generators
     class InstallGenerator < Rails::Generators::Base
       source_root File.expand_path('../templates', __FILE__)
-      class_option :theme, type: :string, default: '', desc: "Use default or 'material' theme"
-
+
```
Comment (monkeyci) [inline lib/generators/activeadmin_addons/install/install_generator.rb:5]: Trailing whitespace detected.

File: lib/generators/activeadmin_addons/assets/assets_generator.rb [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+          to_add = "$am-theme-primary: #342e48;\n"
+          to_add += "@import 'activeadmin_addons/all';\n"
+          to_add + "@import 'activeadmin_addons/material';\n"
+        else
+          "@import 'activeadmin_addons/all';\n"
+        end
+      end
+    end
+  end
+end
```
Comment (monkeyci) [inline lib/generators/activeadmin_addons/assets/assets_generator.rb:53]: Final newline missing.

File: app/assets/stylesheets/activeadmin_addons/imports/webpack/jquery-datepicker.scss [hunk 18]
```diff
@@ -0,0 +1 @@
+@import "jquery-datetimepicker/build/jquery.datetimepicker.min.css";
```
Comment (monkeyci) [inline app/assets/stylesheets/activeadmin_addons/imports/webpack/jquery-datepicker.scss:1]: Expected single quotes   string-quotes

File: app/assets/stylesheets/activeadmin_addons/all.scss [hunk 19]
```diff
@@ -16,4 +16,4 @@ li.error {
   .select2-selection {
     border-color: $error-color;
   }
-}
+}
```
Comment (monkeyci) [inline app/assets/stylesheets/activeadmin_addons/all.scss:19]: Unexpected missing end-of-source newline   no-missing-end-of-source-newline

File: lib/generators/activeadmin_addons/assets/assets_generator.rb [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+          inject_into_file("#{file_path}.js", js_assets, after: reference)
+        end
+      end
+
+      def add_stylesheets
+        file_path = 'app/assets/stylesheets/active_admin'
+
+        begin
+          prepend_file("#{file_path}.scss", css_assets)
+        rescue
```
Comment (monkeyci) [inline lib/generators/activeadmin_addons/assets/assets_generator.rb:23]: Avoid rescuing without specifying an error class.

File: lib/generators/activeadmin_addons/assets/assets_generator.rb [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+    class AssetsGenerator < Rails::Generators::Base
+      class_option :theme, type: :string, default: '', desc: "Use default or 'material' theme"
+
+      def add_javascripts
+        file_path = 'app/assets/javascripts/active_admin'
+        reference = "#= require active_admin/base\n"
+
+        begin
+          inject_into_file("#{file_path}.js.coffee", coffee_assets, after: reference)
+        rescue
```
Comment (monkeyci) [inline lib/generators/activeadmin_addons/assets/assets_generator.rb:12]: Avoid rescuing without specifying an error class.

File: app/javascript/activeadmin_addons/all.js [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+import './config';
+import './inputs/select2';
+import './inputs/search-select';
+import './inputs/nested-select';
+import './inputs/tags';
+import './inputs/selected-list';
+import './inputs/date-time-picker';
+import './inputs/color-picker';
+import './addons/toggle_bool';
+import './addons/interactive_select_tag';
```
Comment (monkeyci) [inline app/javascript/activeadmin_addons/all.js:14]: Newline required at end of file but not found    eol-last

File: app/javascript/activeadmin_addons/all.js [hunk 23]
```diff
@@ -0,0 +1,14 @@
+import 'select2';
```
Comment (monkeyci) [inline app/javascript/activeadmin_addons/all.js:1]: Definition for rule 'vue/max-len' was not found  vue/max-len

File: app/assets/stylesheets/activeadmin_addons/imports/jquery-datepicker.scss [hunk 24]
```diff
@@ -0,0 +1 @@
+@import 'jquery.xdan.datetimepicker';
```
Comment (monkeyci) [inline app/assets/stylesheets/activeadmin_addons/imports/jquery-datepicker.scss:1]: Unexpected missing end-of-source newline   no-missing-end-of-source-newline

File: app/assets/javascripts/activeadmin_addons/all.js [hunk 25]
```diff
@@ -1,14 +1,752 @@
...
+            selectTag.addClass("select-container-hidden");
+            tagContainer.removeClass("interactive-tag-hidden");
+          },
+          type: "PATCH"
+        });
+      }
+    });
+  };
+  $(initializer$9);
+  $(document).on("turbolinks:load", initializer$9);
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/all.js:751]: Strings must use singlequote                                           quotes

File: app/assets/javascripts/activeadmin_addons/all.js [hunk 26]
```diff
@@ -1,14 +1,752 @@
...
+            statusTag.removeClass(oldValue);
+            statusTag.addClass(newValue);
+            tagContainer.data("value", newValue);
+            selectTag.data("value", newValue);
+          },
+          complete: function() {
+            selectTag.addClass("select-container-hidden");
+            tagContainer.removeClass("interactive-tag-hidden");
+          },
+          type: "PATCH"
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/all.js:745]: Strings must use singlequote                                           quotes<br>Missing trailing comma                                                 comma-dangle

File: rollup.config.js [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+    commonjs(),
+    babel(),
+    uglify(uglifyOptions),
+  ],
+  // Use client's yarn dependencies instead of bundling everything
+  external: [
+    'jquery-datetimepicker',
+    'select2',
+  ],
+};
```
Comment (monkeyci) [inline rollup.config.js:42]: Newline required at end of file but not found    eol-last

**Comment by filipkis** [comment]
Thanks @gmq for this PR. Any update on this? What's still needed for it to get merged and released?

**Comment by nicolasrsande** [comment]
+1 Any updates on this? Can it be merged? I Am willing to help if anything is still needed.

**Comment by hgeorgilas** [comment]
Is this going to be merged soon? 
Thanks

**Comment by guiferrpereira** [comment]
excellent work here. can you tell when this will be merged?

**Comment by gmq** [comment]
I can't give a specific timeline, it still needs to be reviewed by the team and time is scarce these days. But it should be sooner rather than later!

**Review by rjherrera — APPROVED** [review]

**Comment by kwent** [comment]
Looking forward to this ! Much needed !

**Comment by kwent** [comment]
@gmq @rjherrera Thanks for the amazing work! Looks like it's very close to the finish line. How can we help ? Regards

**Review by ldlsegovia — APPROVED** [review]

**Comment by dkniffin** [comment]
@gmq Great work on this! I'd love to start using it. What are we waiting on to merge? A review from @hidalgoegz ?

**Comment by qblanc** [comment]
Very happy to find this issue, unable to use the gem with a Rails 6.0 app because of this issue. Thanks for the great work you're doing, can't wait to test it !

**Comment by gmq** [comment]
We're ready to merge to master!

We'd appreciate any bug reports or other feedback about this so if you'd like to test it before we make a new release, please add the following to your Gemfile:

```gem 'activeadmin_addons', git: 'https://github.com/platanus/activeadmin_addons'```

**Comment by agrobbin** [comment]
@gmq this is great! Out of curiosity, are there also plans to remove the hard dependency on `sassc` and `sassc-rails` from the [`gemspec`](https://github.com/platanus/activeadmin_addons/blob/db8ee1394f06243785983410406da22c19b0cd8f/activeadmin_addons.gemspec#L24-L25)? Active Admin [just did that](https://github.com/activeadmin/activeadmin/pull/6584) as well.

**Comment by djalmaaraujo** [comment]
@gmq I can't make it work on activeadmin latest version. Just says activeadmin_addons/all cannot be found. yarn was added, etc. Javascript file is imported but SCSS does not work even with sass-rails on the gemfile.

**Comment by jasonperrone** [comment]
I'm concerned about the comment on this change "if ActiveAdmin was installed in that mode". What about the situation where ActiveAdmin was not installed in that mode, but was later changed to use webpacker?

**Comment by matiasgarcia** [comment]
Same as @djalmaaraujo I can't make it work.

Node module is:

Screenshot from 2021-03-08 11-46-36

I tried doing `@import "activeadmin_addons"` instead of `@import "activeadmin_addons/all"` and now I get this error:

```
ERROR in ./app/javascript/stylesheets/active_admin.scss (./node_modules/css-loader/dist/cjs.js??ref--6-1!./node_modules/postcss-loader/src??ref--6-2!./node_modules/sass-loader/dist/cjs.js??ref--6-3!./app/javascript/stylesheets/active_admin.scss)
Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
SassError: Invalid CSS after "i": expected 1 selector or at-rule, was "import 'select2';"
        on line 1 of node_modules/activeadmin_addons/src/all.js
        from line 13 of /home/mgarcia/projects/withdrawals/app/javascript/stylesheets/active_admin.scss
>> import 'select2';
```

Which fails in the first line of `all.scss` that does:

```
@import 'select2';
```

**Comment by gmq** [comment]
> @gmq I can't make it work on activeadmin latest version. Just says activeadmin_addons/all cannot be found. yarn was added, etc. Javascript file is imported but SCSS does not work even with sass-rails on the gemfile.

> Same as @djalmaaraujo I can't make it work.

I just merged a fix by @ParamagicDev for the documentation. The proper way to import the scss is to use `@import 'activeadmin_addons/src/stylesheets/all';`

> I'm concerned about the comment on this change "if ActiveAdmin was installed in that mode". What about the situation where ActiveAdmin was not installed in that mode, but was later changed to use webpacker?

It checks `config.use_webpacker` in the Active Admin initializer so the addons generator should make the appropriate changes as long as it's enabled there.

**Pull request merged**